opengrok-indexer/src/main/java/org/opengrok/indexer/history/FileHistoryCache.java

```diff
@@ -103,38 +103,32 @@ private void doFileHistory(String filename, List<HistoryEntry> historyEntries,
             Repository repository, File srcFile, File root, boolean renamed)
             throws HistoryException {
 
-        History hist = null;
+        File file = new File(root, filename);
+        // Only store directory history for the top-level directory.
+        if (file.isDirectory() && !filename.equals(repository.getDirectoryName())) {
+            LOGGER.log(Level.FINE, "Not storing history cache for {0}", file);
+            return;
+        }
 
         /*
          * If the file was renamed (in the changesets that are being indexed),
          * its history is not stored in the historyEntries so it needs to be acquired
          * directly from the repository.
-         * This ensures that complete history of the file (across renames)
-         * will be saved.
+         * This ensures that complete history of the file (across renames) will be saved.
          */
+        History hist;
         if (renamed) {
             hist = repository.getHistory(srcFile);
+        } else {
+            hist = new History(historyEntries);
         }
 
-        File file = new File(root, filename);
-
-        if (hist == null) {
-            hist = new History();
-
-            // File based history cache does not store files for individual
-            // changesets so strip them unless it is history for the repository.
-            for (HistoryEntry ent : historyEntries) {
-                if (file.isDirectory() && filename.equals(repository.getDirectoryName())) {
-                    ent.stripTags();
-                } else {
-                    ent.strip();
-                }
-            }
-
-            // add all history entries
-            hist.setHistoryEntries(historyEntries);
-        } else {
-            for (HistoryEntry ent : hist.getHistoryEntries()) {
+        // File based history cache does not store files for individual
+        // changesets so strip them unless it is history for the repository.
+        for (HistoryEntry ent : hist.getHistoryEntries()) {
```

Comment:
hist could be null when `hist = repository.getHistory(srcFile)` returns null which was handled in the previous code

Reply:
If it's part of the contract that it can return `null`, then it would be better to rewrite it to return `Optional`.

Reply:
Sounds to me like a complication. If the history cannot be retrieve, the `HistoryException` should be thrown.

Reply:
There's no word about `null` on https://github.com/oracle/opengrok/blob/ecb1ce82e129709e5bdf11c88db805a26b2bd7ef/opengrok-indexer/src/main/java/org/opengrok/indexer/history/Repository.java#L92-L98
and looking at a few repository implementations they do throw the `HistoryException`, e.g. https://github.com/oracle/opengrok/blob/ecb1ce82e129709e5bdf11c88db805a26b2bd7ef/opengrok-indexer/src/main/java/org/opengrok/indexer/history/MercurialHistoryParser.java#L76-L101
I wonder if it is possible to enforce the class methods not to return null. The `@NotNull` annotation seems to be meant for something else ? Otherwise, provide a wrapper to check for `null` and throw the exception.

Reply:
https://github.com/JetBrains/java-annotations has `@NotNull` that might work.

Reply:
It's ok, I didn't see this conversation https://github.com/oracle/opengrok/pull/3248#discussion_r494812868